File: Subsolar-Point.swift
//: ## Playground - Compute the Real-Time Subsolar Point Coordinates
//: ###### Based on the algorithm used in ISS Real-Time Tracker 3D.
//: ###### Created by Michael Stebel on 02/26/2021.
//: ###### Copyright © 2021-2025 Michael Stebel Consulting, LLC. All rights reserved.
import Foundation
import PlaygroundSupport

/// Variables and constants used by this playground
struct Constants {
    static let coordinatesStringFormat          = "%3d° %02d' %@  %3d° %02d' %@"
    static let degreesLongitudePerHour: Double  = 15
    static let degreesToRadians: Double         = 0.01745329252
    static let earthTiltInDegrees: Double       = 23.43715
    static let earthTiltInRadians: Double       = 0.40905543478
    static let julianDateForJan011970At0000GMT  = 2440587.5
    static let noonTime: Double                 = 12
    static let numberOfDaysInACentury: Double   = 36525
    static let numberOfDaysInAYear: Double      = 365
    static let numberOfHoursInADay: Double      = 24
    static let numberOfMinutesInADay: Double    = 1440
    static let numberOfMinutesInAYear: Double   = numberOfDaysInAYear * numberOfMinutesInADay
    static let numberOfMinutesInAnHour: Double  = 60
    static let numberOfSecondsInADay: Double    = 86400
    static let numberOfSecondsInAMinute: Double = 60
    static let numberOfSecondsInAYear: Double   = numberOfSecondsInADay * numberOfDaysInAYear
    static let numberOfSecondsInAnHour: Double  = 3600
    static let oneEightyDegrees: Double         = 180
    static let radiansToDegrees: Double         = 57.2957795131
    static let threeSixtyDegrees: Double        = 360
}

/// Format converters for coordinates
struct CoordinateConversions {
    /// Convert coordinates from decimal to degrees, minutes, seconds, and direction
    ///
    /// This is a format conversion only.
    /// - Parameters:
    ///   - latitude: Latitude as a Double
    ///   - longitude: Longitude as a Double
    ///   - format: The format to use in the conversion as a String
    /// - Returns: The coordinates string in deg min sec format
    static func decimalCoordinatesToDegMinSec(latitude: Double, longitude: Double, format: String) -> String {
        var latSeconds  = Int(latitude * Double(Constants.numberOfSecondsInAnHour))
        let latDegrees  = latSeconds / Int(Constants.numberOfSecondsInAnHour)
        latSeconds      = abs(latSeconds % Int(Constants.numberOfSecondsInAnHour))
        let latMinutes  = latSeconds / Int(Constants.numberOfSecondsInAMinute)
        latSeconds %= Int(Constants.numberOfSecondsInAMinute)

        var lonSeconds = Int(longitude * Double(Constants.numberOfSecondsInAnHour))
        let lonDegrees = lonSeconds / Int(Constants.numberOfSecondsInAnHour)
        lonSeconds     = abs(lonSeconds % Int(Constants.numberOfSecondsInAnHour))
        let lonMinutes = lonSeconds / Int(Constants.numberOfSecondsInAMinute)
        lonSeconds %= Int(Constants.numberOfSecondsInAMinute)

        return String(format: format, abs(latDegrees), latMinutes, { latDegrees >= 0 ? "North" : "South" }(), abs(lonDegrees), lonMinutes, { lonDegrees >= 0 ? "East" : "West" }())
    }

    /// Convert coordinates from degrees, minutes, seconds, and direction to decimal
    ///
    /// This is a format conversion only.
    /// - Parameters:
    ///   - degrees: Degrees as a Double
    ///   - minutes: Minutes as a Double
    ///   - seconds: Seconds as a Double
    ///   - direction: Direction as a String (either "N" or "S")
    /// - Returns: Decimal representation of coordinates as a Double
    static func degMinSecCoordinatesToDecimal(degrees: Double, minutes: Double, seconds: Double, direction: String) -> Double {
        let sign = (direction == "S" || direction == "W") ? -1.0 : 1.0

        return (degrees + (minutes + seconds / Double(Constants.numberOfSecondsInAMinute)) / Double(Constants.numberOfSecondsInAMinute)) * sign
    }
}

struct Astrocalculations {
    /// Convert a given Gregorian date to a Julian date
    /// - Parameter date: Gregorian date to convert as a Date
    /// - Returns: Julian date as a Double
    static func jDFromDate(date: Date) -> Double {
        let jD = Constants.julianDateForJan011970At0000GMT + date.timeIntervalSince1970 / Double(Constants.numberOfSecondsInADay)
        
        return jD
    }
    
    /// Convert a given Julian date to a Gregorian date
    /// - Parameter jd: Julian date as a Double
    /// - Returns: Standard date as a NSDate
    static func dateFromJd(jd: Double) -> NSDate {
        let gD = NSDate(timeIntervalSince1970: (jd - Constants.julianDateForJan011970At0000GMT) * Double(Constants.numberOfSecondsInADay))
        
        return gD
    }
    
    /// Calculate the Julian century since Jan-1-2000
    /// - Parameter date: Gregorian date as a Date
    /// - Returns: Julian century as a Double
    static func julianCenturySinceJan2000(date: Date) -> Double {
        let jc = (jDFromDate(date: date) - 2451545) / Double(Constants.numberOfDaysInACentury)
        
        return jc
    }
    
    /// Calculate the orbital eccentricity of the Earth
    ///
    /// The orbital eccentricity of an astronomical object is a dimensionless parameter that determines the amount by which its orbit around another body deviates from a perfect circle.
    /// - Parameter t: Julian century as a Double
    /// - Returns: Eccentricity as a Double
    static func orbitEccentricityOfEarth(t: Double) -> Double {
        let ecc = 0.016708634 - t * (0.000042037 + 0.0000001267 * t)
        
        return ecc
    }
    
    /// Calculate the Mean Anomaly of the Sun for a given date
    ///
    /// The mean anomaly is the angle between lines drawn from the Sun to the perihelion and to a point moving in the orbit at a uniform rate corresponding to the period of revolution of the planet.
    /// If the orbit of the planet were a perfect circle, then the planet as seen from the Sun would move along its orbit at a fixed speed.
    /// Then it would be simple to calculate its position (and also the position of the Sun as seen from the planet).
    /// The position that the planet would have relative to its perihelion if the orbit of the planet were a circle is called the mean anomaly.
    /// - Parameter t: Julian century as a Double
    /// - Returns: The mean anomaly as a Double
    static func meanAnomaly(t: Double) -> Double {
        let m = 357.52911 + t * 35999.05029 - t * 0.0001537
        
        return m
    }
    
    /// Calculate the Equation of Time for a given date
    ///
    /// The Equation of Time (EOT) is a formula used to convert between solar time and clock time, accounting for the Earth’s elliptical orbit around the Sun and its axial tilt.
    /// Essentially, the Earth does not move perfectly smoothly in a perfectly circular orbit, so the EOT adjusts for that.
    /// - Parameter date: A date as a Date type
    /// - Returns: Equation of time in minutes as a Double
    static func equationOfTime(for date: Date) -> Double {
        let t = julianCenturySinceJan2000(date: date)
        let meanLongitudeSunRadians = geometricMeanLongitudeOfSunAtCurrentTime(t: t) * Constants.degreesToRadians
        let meanAnomalySunRadians = meanAnomaly(t: t) * Constants.degreesToRadians
        let eccentricity = orbitEccentricityOfEarth(t: t)
        let obliquity = 0.0430264916545165
        
        let term1 = obliquity * sin(2 * meanLongitudeSunRadians)
        let term2 = 2 * eccentricity * sin(meanAnomalySunRadians)
        let term3 = 4 * eccentricity * obliquity * sin(meanAnomalySunRadians) * cos(2 * meanLongitudeSunRadians)
        let term4 = 0.5 * obliquity * obliquity * sin(4 * meanLongitudeSunRadians)
        let term5 = 1.25 * eccentricity * eccentricity * sin(2 * meanAnomalySunRadians)
        
        let equationOfTime = 4 * (term1 - term2 + term3 - term4 - term5) * Constants.radiansToDegrees
        
        return equationOfTime
    }
    
    /// Calculate the Sun Equation of Center
    ///
    /// The orbits of the planets are not perfect circles, but rather ellipses, so the speed of the planet in its orbit varies, and therefore the apparent speed of the Sun along the ecliptic also varies throughout the planet's year.
    /// The true anomaly is the angular distance of the planet from the perihelion of the planet, as seen from the Sun. For a circular orbit, the mean anomaly and the true anomaly are the same.
    /// The difference between the true anomaly and the mean anomaly is called the Equation of Center.
    /// - Parameter t: Julian century as a Double
    /// - Returns: The Sun equation of Center in radians
    static func sunEquationOfCenter(t: Double) -> Double {
        let m = meanAnomaly(t: t)
        let sEOC = sin(m * Double(Constants.degreesToRadians)) * (1.914602 - t * (0.004817 + 0.000014 * t)) + sin(2 * m * Double(Constants.degreesToRadians)) * (0.019993 - 0.000101 * t) + sin(3 * m * Double(Constants.degreesToRadians)) * 0.000289
        
        return sEOC
    }
    
    /// Calculate the Geometric Mean Longitude of the Sun
    ///
    /// The mean longitude of the Sun, corrected for the aberration of light. Mean longitude is the ecliptic longitude at which an orbiting body could be found if its orbit were circular and free of perturbations.
    /// While nominally a simple longitude, in practice the mean longitude does not correspond to any one physical angle.
    /// - Parameter t: Julian century as a Double
    /// - Returns: The geometric mean longitude in degrees as a Double
    static func geometricMeanLongitudeOfSunAtCurrentTime(t: Double) -> Double {
        let sunGeometricMeanLongitude = Double(280.46646 + t * 36000.76983 + t * t * 0.0003032).truncatingRemainder(dividingBy: Constants.threeSixtyDegrees)
        
        return sunGeometricMeanLongitude
    }
    
    /// Calculate the exact current latitude of the Sun for a given date and time
    ///
    /// Based on the geometric mean longitude of the Sun.
    /// - Parameter date: A date as a Date type
    /// - Returns: Latitude in degrees as a Double
    static func latitudeOfSun(for date: Date) -> Double {
        let jC = julianCenturySinceJan2000(date: date)
        
        let geomMeanLongitude = geometricMeanLongitudeOfSunAtCurrentTime(t: jC)
        let sunTrueLongitude = geomMeanLongitude + Double(sunEquationOfCenter(t: jC))
        let latitudeOfSun = asin(sin(sunTrueLongitude * Constants.degreesToRadians) * sin(Constants.earthTiltInRadians))
        let sunTrueLatitudeInDegrees = latitudeOfSun * Constants.radiansToDegrees
        
        return sunTrueLatitudeInDegrees
    }
    
    /// Calculate the exact current longitude of the Sun for a given date and time
    ///
    /// This is an original algorithm that I created to calculate the subsolar point longitude, based on a given date and the equation of time.
    /// - Parameter date: A date as a Date type
    /// - Returns: The subsolar longitude as a Double
    static func subSolarLongitudeAtCurrentTimeUTC(for date: Date) -> Double {
        let utcCalendar = Calendar(identifier: .gregorian)
        let utcComponents = utcCalendar.dateComponents(in: TimeZone(secondsFromGMT: 0)!, from: date)

        guard let hour = utcComponents.hour,
              let minute = utcComponents.minute,
              let second = utcComponents.second else {
            return 0.0
        }

        // Convert UTC time to decimal hours
        let utcDecimalHours = Double(hour) + Double(minute) / Constants.numberOfMinutesInAnHour + Double(second) / Constants.numberOfSecondsInAnHour

        // Get the Equation of Time in minutes
        let eotMinutes = equationOfTime(for: date)
        let eotHours = eotMinutes / Constants.numberOfMinutesInAnHour

        // Subsolar longitude with Equation of Time correction
        var longitude = (12.0 - utcDecimalHours - eotHours) * Constants.degreesLongitudePerHour

        // Normalize to [-180°, 180°]
        if longitude > 180 {
            longitude -= 360
        } else if longitude < -180 {
            longitude += 360
        }

        return longitude
    }
    
    /// Get the subsolar coordinates at the current date and time
    ///
    /// The subsolar point is the position on Earth where the Sun is at the zenith.
    /// - Returns: The subsolar coordinates as a tuple of Doubles
    static func getSubSolarCoordinates() -> (latitude: Float, longitude: Float) {
        let now = Date()
        let lat = Float(latitudeOfSun(for: now))
        let lon = Float(subSolarLongitudeAtCurrentTimeUTC(for: now))
                        
        return (lat, lon)
    }
}

//: ### Run it in a loop
let secsToDelay: UInt32 = 5

while true {
    let now           = Date()
    let subsolarPoint = Astrocalculations.getSubSolarCoordinates()
    let lat           = Double(subsolarPoint.latitude)
    let lon           = Double(subsolarPoint.longitude)
    let coordinates   = CoordinateConversions.decimalCoordinatesToDegMinSec(latitude: lat, longitude: lon, format: Constants.coordinatesStringFormat)
    print("The subsolar point is now at: \(coordinates) (at \(now))")
    sleep(secsToDelay)
}

File: SubSolarPoint.py
import math
import time
from datetime import datetime, timezone, timedelta

class Constants:
    coordinates_string_format = "{:3d}° {:02d}' {}  {:3d}° {:02d}' {}"
    degrees_longitude_per_hour = 15
    degrees_to_radians = math.pi / 180
    earth_tilt_in_radians = 0.40905543478
    julian_date_for_jan_01_1970 = 2440587.5
    noon_time = 12
    number_of_days_in_a_century = 36525
    number_of_days_in_a_year = 365
    number_of_minutes_in_an_hour = 60
    number_of_minutes_in_a_day = 1440
    number_of_seconds_in_a_day = 86400
    number_of_seconds_in_a_year = number_of_seconds_in_a_day * number_of_days_in_a_year
    number_of_seconds_in_an_hour = 3600
    radians_to_degrees = 180 / math.pi
    three_sixty_degrees = 360
    one_eighty_degrees = 180
    number_of_hours_in_a_day = 24

def decimal_to_deg_min_sec(lat, lon):
    def convert(coord, pos_dir, neg_dir):
        total_seconds = int(coord * Constants.number_of_seconds_in_an_hour)
        degrees = total_seconds // Constants.number_of_seconds_in_an_hour
        seconds = abs(total_seconds % Constants.number_of_seconds_in_an_hour)
        minutes = seconds // Constants.number_of_minutes_in_an_hour
        direction = pos_dir if degrees >= 0 else neg_dir
        return abs(degrees), minutes, direction

    lat_d, lat_m, lat_dir = convert(lat, "North", "South")
    lon_d, lon_m, lon_dir = convert(lon, "East", "West")

    return Constants.coordinates_string_format.format(lat_d, lat_m, lat_dir, lon_d, lon_m, lon_dir)

def julian_date_from_datetime(dt):
    unix_time = dt.replace(tzinfo=timezone.utc).timestamp()
    return Constants.julian_date_for_jan_01_1970 + unix_time / Constants.number_of_seconds_in_a_day

def julian_century_since_2000(dt):
    return (julian_date_from_datetime(dt) - 2451545.0) / Constants.number_of_days_in_a_century

def geometric_mean_longitude_sun(t):
    return (280.46646 + t * 36000.76983 + t * t * 0.0003032) % Constants.three_sixty_degrees

def mean_anomaly(t):
    return 357.52911 + t * (35999.05029 - 0.0001537)

def eccentricity_of_earth_orbit(t):
    return 0.016708634 - t * (0.000042037 + 0.0000001267 * t)

def sun_equation_of_center(t):
    m = mean_anomaly(t)
    m_rad = m * Constants.degrees_to_radians
    return (math.sin(m_rad) * (1.914602 - t * (0.004817 + 0.000014 * t)) +
            math.sin(2 * m_rad) * (0.019993 - 0.000101 * t) +
            math.sin(3 * m_rad) * 0.000289)

def latitude_of_sun(dt):
    t = julian_century_since_2000(dt)
    geom_mean_long = geometric_mean_longitude_sun(t)
    true_long = geom_mean_long + sun_equation_of_center(t)
    true_long_rad = true_long * Constants.degrees_to_radians
    lat_rad = math.asin(math.sin(true_long_rad) * math.sin(Constants.earth_tilt_in_radians))
    return lat_rad * Constants.radians_to_degrees

def equation_of_time(dt):
    t = julian_century_since_2000(dt)
    L0 = geometric_mean_longitude_sun(t) * Constants.degrees_to_radians
    e = eccentricity_of_earth_orbit(t)
    M = mean_anomaly(t) * Constants.degrees_to_radians
    obliquity = 0.0430264916545165  # ~23.44 degrees in radians

    term1 = obliquity * math.sin(2 * L0)
    term2 = 2 * e * math.sin(M)
    term3 = 4 * e * obliquity * math.sin(M) * math.cos(2 * L0)
    term4 = 0.5 * obliquity * obliquity * math.sin(4 * L0)
    term5 = 1.25 * e * e * math.sin(2 * M)

    eot = 4 * (term1 - term2 + term3 - term4 - term5) * Constants.radians_to_degrees
    return eot

def subsolar_longitude(dt):
    eot = equation_of_time(dt)
    local_time = dt.astimezone()
    hour = local_time.hour + local_time.minute / 60.0
    seconds_from_gmt = local_time.utcoffset().total_seconds()

    if seconds_from_gmt <= 0:
        time_correction = 1
        day_correction = 0
    else:
        time_correction = -1
        day_correction = -Constants.number_of_hours_in_a_day

    gmt_hour = (hour - seconds_from_gmt / Constants.number_of_seconds_in_an_hour -
                time_correction * Constants.number_of_hours_in_a_day) % Constants.number_of_hours_in_a_day

    noon_delta = min(Constants.noon_time - gmt_hour - eot / Constants.number_of_minutes_in_an_hour + day_correction, 12.0)
    subsolar_lon = noon_delta * Constants.degrees_longitude_per_hour

    if subsolar_lon < -Constants.one_eighty_degrees and gmt_hour <= Constants.noon_time:
        lon_correction = Constants.one_eighty_degrees
    elif subsolar_lon <= -Constants.one_eighty_degrees and gmt_hour >= Constants.noon_time:
        lon_correction = 0
    elif gmt_hour >= Constants.number_of_hours_in_a_day:
        lon_correction = Constants.one_eighty_degrees
    else:
        lon_correction = 0

    return subsolar_lon + lon_correction

def get_subsolar_coordinates():
    now = datetime.now()
    lat = latitude_of_sun(now)
    lon = subsolar_longitude(now)
    return lat, lon

# Main loop
while True:
    now = datetime.now()
    lat, lon = get_subsolar_coordinates()
    coords = decimal_to_deg_min_sec(lat, lon)
    print(f"The subsolar point is now at: {coords} (at {now})")
    time.sleep(5)
